### PR TITLE
Improvement of Dual Unit Display

### PR DESCRIPTION
**Pull request by ao3742**
Hi,

This version has the solution for femr-136 - dual unit display issue.

Following files were modified for dual unit display .

        modified:   app/femr/business/services/system/PatientService.java
        modified:   app/femr/business/services/system/SearchService.java
        modified:   app/femr/business/services/system/VitalService.java
        modified:   app/femr/common/models/PatientItem.java
        modified:   app/femr/ui/views/medical/listVitals.scala.html
        modified:   app/femr/ui/views/partials/patientOverview.scala.html
        modified:   app/femr/util/calculations/LocaleUnitConverter.java

I've added comments for my changes, my comment will start with femr-136 - dual unit display

Thank You.

### CODE REVIEW

**Comment by kevinzurek** [comment]
Looking good - 

I had the Celsius and Fahrenheit temperature values become interchanged:
cf


To reproduce:
1) Log in with administrator account, make sure the Metric System Option is turned off
2) Create new patient and set temperature = 98.7F
3) Turn on the Metric System Option
4) The temperature in the vital section of Medical looks like the screenshot above

**Comment by AJSclyn** [comment]
Kevin - 

It looks to be fixed.  I could not reproduce from my end.

File: app/femr/ui/views/partials/patientOverview.scala.html [hunk 1]
```diff
@@ -23,8 +23,13 @@
...
-        <label>Weight: </label> <span>@outputWeightOrNA(String.valueOf(patient.getWeight), settings.isMetric)</span>
+        @if(settings.isMetric) {
+            <label>Height: </label> <span>@outputHeightOrNA(String.valueOf(patient.getHeightFeet), String.valueOf(patient.getHeightInches), settings.isMetric) / @outputHeightOrNA(String.valueOf(patient.getHeightFeetDual), String.valueOf(patient.getHeightInchesDual), !settings.isMetric)</span>
+            <label>Weight: </label> <span>@outputWeightOrNA(String.valueOf(patient.getWeight), settings.isMetric) / @outputWeightOrNA(String.valueOf(patient.getWeightDual), !settings.isMetric)</span>
+        } else {
+            <label>Height: </label> <span>@outputHeightOrNA(String.valueOf(patient.getHeightFeet), String.valueOf(patient.getHeightInches), settings.isMetric) / @outputHeightOrNA(String.valueOf(patient.getHeightFeetDual), String.valueOf(patient.getHeightInchesDual), !settings.isMetric)</span>
+            <label>Weight: </label> <span>@outputWeightOrNA(String.valueOf(patient.getWeight), settings.isMetric) / @outputWeightOrNA(String.valueOf(patient.getWeightDual), !settings.isMetric)</span>
+        }
```
Comment (kevinzurek) [inline app/femr/ui/views/partials/patientOverview.scala.html]: I don't think the if statement is needed here - each condition has the same code?

**Comment by kevinzurek** [comment]
Yep, it works :+1:

**Comment by ao3742** [comment]
Hi Kevin, yes I  agree, if condition is not necessary. I will remove the if condition and resubmit app/femr/ui/views/partials/patientOverview.scala.html only tonight.

File: app/femr/util/calculations/LocaleUnitConverter.java [hunk 2]
```diff
@@ -18,6 +18,52 @@
...
+
+        for (int dateIndex = 0; dateIndex < vitalMap.getDateListChronological().size(); dateIndex++) {
+
+            // Get imperial temperature(F)
+            String tempC = vitalMap.get("temperature", vitalMap.getDate(dateIndex));
+
+            // If temp is not null convert to metric(C) and store in map as temperature
+            if (tempC != null) {
+                vitalMap.put("temperature", vitalMap.getDate(dateIndex), getCelcius(tempC));  // temperature is in Fahrenheit
+                vitalMap.put("temperatureC", vitalMap.getDate(dateIndex), tempC);   // added for femr-136 - dual unit display
```
Comment (kevinzurek) [inline app/femr/util/calculations/LocaleUnitConverter.java]: Is the C supposed to stand for Celsius or would this make more sense as "temperatureF" or "temperatureFahrenheit"?

**Comment by kevinzurek** [comment]
Thanks Vivek, code seems to work well - just trying to understand the naming conventions :+1:  :)

**Comment by ao3742** [comment]
Hi Kevin,

vitalMap.put("temperature", vitalMap.getDate(dateIndex), getCelcius(tempC));   was already in the toMetric function. 

I only added vitalMap.put("temperatureC", vitalMap.getDate(dateIndex), tempC);   

If you want I can rename temperatureC to temperatureCelsius.

Next, if I changed temperature to temperatureFahrenheit I have to make changes in scala.html file also that's why I did not change.

Please let me know if you want me to rename temperatureC to temperatureCelsius

Thanks, Vivek

**Comment by kevinzurek** [comment]
Don't worry about making changes in too many files - more the better if it means consistent naming.

File: app/femr/ui/views/medical/listVitals.scala.html [hunk 3]
```diff
@@ -41,11 +41,10 @@
...
             <th>T</th>
             @for(dateIndex <- 1 to vitalMap.getDateListChronological.size) {
-                <td>@outputTemperatureOrNA(
-                        vitalMap.get("temperature", vitalMap.getDate(dateIndex - 1)),
-                        viewModel.getSettings.isMetric,
-                        ""
-                )</td>
+                <td>
+                     @outputTemperatureOrNA(vitalMap.get("temperatureFahernheit", vitalMap.getDate(dateIndex - 1)), viewModel.getSettings.isMetric, "")
+                    /@outputTemperatureOrNA(vitalMap.get("temperatureCelsius", vitalMap.getDate(dateIndex - 1)), !viewModel.getSettings.isMetric, "")
```
Comment (kevinzurek) [inline app/femr/ui/views/medical/listVitals.scala.html]: If I turn on the metric system option, then temperatureFahrenheit stores the celsius value

File: app/femr/ui/views/medical/listVitals.scala.html [hunk 4]
```diff
@@ -94,6 +99,11 @@
                         vitalMap.get("heightInches", vitalMap.getDate(dateIndex - 1)),
                         viewModel.getSettings.isMetric,
                         ""
+                    ) /@outputHeightOrNA(
+                        vitalMap.get("heightMeters", vitalMap.getDate(dateIndex - 1)),
+                        vitalMap.get("heightCm", vitalMap.getDate(dateIndex - 1)),
+                        !viewModel.getSettings.isMetric,
+                        ""
                     )
```
Comment (kevinzurek) [inline app/femr/ui/views/medical/listVitals.scala.html]: The way you manage this is perfect "heightMeters" always has the height value in meters and "heightFeet" always has the height value in feet

**Comment by kevinzurek** [comment]
:+1:

**Comment by kdunlap** [comment]
Manually merged

https://teamfemr.atlassian.net/browse/FEMR-136

**Pull request closed without merging**